Implement url analyzer API for import onboarding

jeroenpf opened the pull request:
#### Changes proposed in this Pull Request

This change implements the API endpoint for the URL analyzer in the import onboarding flow at `/start/import/capture`

#### Testing instructions

 - Go to `/start/import/capture`.
 - Enter a valid URL and hit enter
 - The scanning step will show - possibly really quick - followed by the preview page.
 - Check if the site address and the platform are correctly displayed in the content below the page title.

Not all functionality has been fully implemented yet. This PR purely focuses on getting the data from the URL analyzer endpoint and making it available in the step components.

Related to #57131

jeroenpf committed:
Add url analyzer endpoint

jeroenpf committed:
Remove goToNextStep prop

jeroenpf committed:
Merge branch 'trunk' into add/gutenboarding-url-analyzer-api

jeroenpf committed:
Pass url data to the preview step

jeroenpf committed:
Move url analyzing state to redux

jeroenpf committed:
Use connectedProps

vishnugopal committed:
Fix all typescript errors

pento commented on client/lib/wpcom-undocumented/lib/undocumented.js:
```diff
@@ -1925,4 +1925,18 @@ Undocumented.prototype.getMatchingAnchorSite = function (
 	);
 };
 
+Undocumented.prototype.analyzeUrl = function ( url, fn ) {
```
Perhaps call this function `ImportsAnalyzeUrl()`? It's usually a good idea to match the function name with the API call.

It needs a docblock, too. 🙂

vishnugopal commented on client/signup/steps/import/capture/index.tsx:
```diff
@@ -13,39 +16,31 @@ const validateUrl = ( url: string ): boolean => {
-		 */
-		setTimeout( () => {
+		analyzeUrl( urlValue ).then( () => {
+			/**
```
Remove this comment?

jeroenpf replied:
@vishnugopal I left it there because the logic here is not complete and Bogdan has added the comment for when this will be finished. I will leave it here for now.

vishnugopal commented on client/signup/steps/import/index.tsx:
```diff
@@ -11,59 +13,60 @@ interface Props {
+	isAnalyzing: boolean;
 }
 
 const MOCK_DATA = {
```
Remove this?

vishnugopal reviewed (commented):
A few bits of unneeded code, but otherwise looks good!

jeroenpf committed:
Show error notification when API request fails

jeroenpf committed:
Rename the analyzeUrl api function

Additionally, the docblock was missing and has been added

jeroenpf committed:
Refactor import step to use ConnectedProps

jeroenpf committed:
Clear the url analyzer api error

When the user starts to type a URL, previous errors need to be cleared.

vishnugopal commented:
@jeroenpf Can we whip up the changes required to address feedback from Gary as well now? Thanks!

vishnugopal commented:
I'm also getting this rather strange error when I try to test:

https://user-images.githubusercontent.com/533/139787210-5e2a2952-c851-4b2d-8e9d-6ef16185577a.mp4

jeroenpf commented:
> I'm also getting this rather strange error when I try to test:
> 
>  CleanShot.2021-11-02.at.10.01.00.mp4

Looking into this now. I am not sure what is causing this.

jeroenpf committed:
Remove unused dependency

The dependecy for the ready step is not used and is also causing an issue that loading the capture page redirects you to the domain step.

jeroenpf commented:
@vishnugopal I've seen the same issue now and it seems to be related to the dependency that was not used. I removed it now: 5cccaf4119ddd314c6aaa01d24c95a6aa534ddc8

jeroenpf committed:
Remove trailing slash if there is no path

If no path is provided for the analyzed URL, we only want to display the hostname.

jeroenpf committed:
Map name of platform to display name

jeroenpf committed:
Fix for going to the wrong view when platform is unknown

When the url analyzer API returns unknown for the platform, we want to show a different next step. This commit ensures that the correct view is shown when we can't determine the platform of the site

jeroenpf committed:
Show the preview image when its done generating

Initially, mShots returns a loading spinner gif when requesting a screenshot for a given URL. This fix keeps checking the mShots url until it returns a screenshot instead of the loading spinner and then displays the image to the user

jeroenpf commented:
@vishnugopal @pento Most of the issues that @pento pointed out have been addressed in this PR.

Bogdan will give it a finishing touch and add a loading spinner for the preview image.  I've only implemented the logic that ensures the preview is shown correctly once it's available but wasn't sure about styling the loading state.

After that, two items are remaining:

- What to show when the site is hosted on WordPress.com - I guess we need to change the content so that it shows instructions for site migration instead? - In any case, this will also require a change in the API endpoint as it does not distinguish between wpcom and self-hosted WordPress sites.
- The language 'is hosted by [platform]' - What should we change this into? Something that works for any platform would be best. Perhaps: ' It looks like [website] is built with [platform] .... `


I've requested a review again from both of you, I'll let you guys decide who gets the honor to do it :-)

vishnugopal commented:
> What to show when the site is hosted on WordPress.com - I guess we need to change the content so that it shows instructions for site migration instead? - In any case, this will also require a change in the API endpoint as it does not distinguish between wpcom and self-hosted WordPress sites. 

Am ok with this moving to a future PR.

ollierozdarz commented:
👋  I think this answers @jeroenpf's questions above. 

<img width="1674" alt="Screen Shot 2021-11-03 at 2 35 00 pm" src="https://user-images.githubusercontent.com/67085494/140010437-fa8ca858-b5e2-4d56-9953-b0ae608c6896.png">

Also, here's the [Figma](https://www.figma.com/file/TAK5FpwTvfbENJDfHc7Aq1/Site-Importing?node-id=1930%3A21705) file.

vishnugopal commented:
> The language 'is hosted by [platform]' - What should we change this into? Something that works for any platform would be best. Perhaps: ' It looks like [website] is built with [platform] .... `

"built with [platform]" sounds great here @jeroenpf and @ollierozdarz concurs!

bogiii committed:
Import ready: integrate spinner component while generating screenshot

bogiii commented:
I've stumbled upon a warning when I load the capture page. We should address it before merging.
```
react-jsx-runtime.development.js:117 Warning: Failed prop type: The prop `stepName` is marked as required in `NavigationLink`, but its value is `undefined`.
    at NavigationLink (app:///./signup/navigation-link/index.jsx:43:5)
    at _class (app:///../packages/i18n-calypso/src/localize.js:36:9)
    at ConnectFunction (app:///../node_modules/react-redux/es/components/connectAdvanced.js:231:68)
    at div
    at ActionButtons (app:///../packages/onboarding/src/action-buttons/index.tsx:29:3)
    at div
    at StepWrapper (app:///./signup/step-wrapper/index.jsx:32:1)
    at _class (app:///../packages/i18n-calypso/src/localize.js:36:9)
    at ImportOnboarding (app:///./signup/steps/import/index.tsx:32:3)
    at ConnectFunction (app:///../node_modules/react-redux/es/components/connectAdvanced.js:231:68)
    at div
    at div
    at div
    at div
    at Signup (app:///./signup/main.jsx:151:5)
    at ConnectFunction (app:///../node_modules/react-redux/es/components/connectAdvanced.js:231:68)
    at div
    at div
    at div
    at Layout (app:///./layout/index.jsx:243:1)
    at ConnectFunction (app:///../node_modules/react-redux/es/components/connectAdvanced.js:231:68)
    at WithCurrentRoute (app:///./components/route/index.js:44:30)
    at MomentProvider (app:///./components/localized-moment/provider.js:27:5)
    at ConnectFunction (app:///../node_modules/react-redux/es/components/connectAdvanced.js:231:68)
    at Provider (app:///../node_modules/react-redux/es/components/Provider.js:18:20)
    at QueryClientProvider (app:///../node_modules/react-query/es/react/QueryClientProvider.js:38:21)
    at RouteProvider (app:///./components/route/index.js:22:3)
    at I18nProvider (app:///../node_modules/@wordpress/react-i18n/build-module/index.js:68:5)
    at LocaleProvider (app:///../packages/i18n-utils/src/locale-context.tsx:19:3)
    at CalypsoI18nProvider (app:///./components/calypso-i18n-provider/index.tsx:20:3)
    at ProviderWrappedLayout (app:///./controller/index.web.js:60:3)
```

jeroenpf committed:
Add stepName attribute to StepWrapper

jeroenpf committed:
Merge branch 'trunk' into add/gutenboarding-url-analyzer-api

bogiii reviewed (approved):
We are good to go! 🚀

tyxla commented:
Hey, @jeroenpf @bogiii @pento @vishnugopal! 

We're in the process of migrating away from the deprecated `wpcom.undocumented`, and I just noticed that this PR adds a new method to it. Would you mind updating to use `wpcom.req.get()` or `wpcom.req.post()` instead?

See p4TIVU-9KX-p2 for more details, and feel free to ping me or anyone from @Automattic/team-calypso-frameworks if you need any help.

Thanks in advance!

jsnajdr commented on client/lib/wpcom-undocumented/lib/undocumented.js:
```diff
@@ -1650,4 +1650,25 @@ Undocumented.prototype.getMatchingAnchorSite = function (
+	return this.wpcom.req.get(
+		{
+			path: '/imports/analyze-url',
+			method: 'GET',
```
By the way the `method` field doesn't need to be there, calling `req.get` already says what HTTP method should be used.

jeroenpf replied:
@jsnajdr Thanks for pointing this out :-) I've already created a new PR to address your feedback: #57815 - It will be merged by tomorrow.
